crates/onnx-ir/src/rank_inference.rs

```
@@ -1074,3 +1075,53 @@ fn one_hot_output_shape(node: &mut Node) {
         elem_type: output_elem,
     });
 }
+
+fn gemm_output_shape(node: &mut Node) {
+    let a_shape = match &node.inputs[0].ty {
+        ArgType::Tensor(tensor) => tensor.shape.clone().expect("Input A should have a shape!"),
+        _ => panic!("Input A should be a tensor!"),
+    };
+    let b_shape = match &node.inputs[1].ty {
+        ArgType::Tensor(tensor) => tensor.shape.clone().expect("Input B should have a shape!"),
+        _ => panic!("Input B should be a tensor!"),
+    };
+    let trans_a = node
+        .attrs
+        .get("transA")
+        .map(|v| v.clone().into_i64())
+        .unwrap_or(0);
+
+    let trans_b = node
+        .attrs
+        .get("transB")
+        .map(|v| v.clone().into_i64())
+        .unwrap_or(0);
+
+    let (m, k_a) = if trans_a != 0 {
+        (a_shape[1], a_shape[0])
+    } else {
+        (a_shape[0], a_shape[1])
+    };
+
+    let (k_b, n) = if trans_b != 0 {
+        (b_shape[1], b_shape[0])
+    } else {
+        (b_shape[0], b_shape[1])
+    };
+
+    if k_a != k_b {
+        panic!(
+            "Gemm: inner dimensions do not match (got {} and {})",
+            k_a, k_b
+        );
+    }
```

This check should not be performed here. The shapes are not always available, so we can let the execution crash during runtime if the provided inputs are not valid.

---

The shape information is needed here though-

```rs
 node.outputs[0].ty = ArgType::Tensor(TensorType {
        rank: 2,
        shape: Some(vec![m, n]),
```

---

Not mandatory no, it can be left as `None` . The shape is not used at compile-time.

Your comment made me realize that you hardcoded the rank as `2`. That's incorrect, it should be the same as the input rank.